File: src/PDFKeeper.WindowsApplication/Common/HelpFile.vb
'******************************************************************************
'* PDFKeeper -- Open Source PDF Document Management
'* Copyright (C) 2009-2019  Robert F. Frasca
'*
'* This file is part of PDFKeeper.
'*
'* PDFKeeper is free software: you can redistribute it and/or modify
'* it under the terms of the GNU General Public License as published by
'* the Free Software Foundation, either version 3 of the License, or
'* (at your option) any later version.
'*
'* PDFKeeper is distributed in the hope that it will be useful,
'* but WITHOUT ANY WARRANTY; without even the implied warranty of
'* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'* GNU General Public License for more details.
'*
'* You should have received a copy of the GNU General Public License
'* along with PDFKeeper.  If not, see <http://www.gnu.org/licenses/>.
'******************************************************************************
Public Class HelpFile
    Private chmFile As String

    Public Sub New()
        chmFile = "PDFKeeper." & CultureInfo.CurrentCulture.ToString & ".chm"
        If IO.File.Exists(chmFile) = False Then
            chmFile = "PDFKeeper.en-US.chm"
        End If
    End Sub

    ''' <summary>
    ''' Returns the name of the help file based on the current culture.
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks>
    ''' If a help file does not exist for the current culture, then the help
    ''' file for en-US will be used.
    ''' </remarks>
    Public ReadOnly Property Name As String
        Get
            Return chmFile
        End Get
    End Property

    ''' <summary>
    ''' Shows the specified topic page in the help file and waits until closed.
    ''' </summary>
    ''' <param name="helpTopic">
    ''' Topic file with extension contained in help file to display.
    ''' </param>
    ''' <remarks></remarks>
    Public Sub ShowAndWait(ByVal helpTopic As String)
        Using htmlHelp As New Process
            htmlHelp.StartInfo.FileName = Path.Combine( _
                Environment.GetFolderPath(Environment.SpecialFolder.Windows), _
                "hh.exe")
            htmlHelp.StartInfo.Arguments = "ms-its:" & chmFile & "::" & helpTopic
            htmlHelp.Start()
            htmlHelp.WaitForExit()
        End Using
    End Sub

    ''' <summary>
    ''' Shows the specified topic in the help file using the Help dialog box.
    ''' </summary>
    ''' <param name="parentControl">
    ''' Parent form or control of the help dialog box. The Me object can be
    ''' specified as the parent control.
    ''' </param>
    ''' <param name="helpTopic">
    ''' Topic file with extension contained in help file to display.
    ''' </param>
    ''' <remarks></remarks>
    Public Sub Show(ByVal parentControl As System.Windows.Forms.Control, _
                    ByVal helpTopic As String)
        Help.ShowHelp(parentControl, chmFile, helpTopic)
    End Sub
End Class

File: src/PDFKeeper.Core/Application/HelpFile.cs
// ****************************************************************************
// * PDFKeeper -- Open Source PDF Document Management
// * Copyright (C) 2009-2025 Robert F. Frasca
// *
// * This file is part of PDFKeeper.
// *
// * PDFKeeper is free software: you can redistribute it and/or modify it
// * under the terms of the GNU General Public License as published by the
// * Free Software Foundation, either version 3 of the License, or (at your
// * option) any later version.
// *
// * PDFKeeper is distributed in the hope that it will be useful, but WITHOUT
// * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
// * more details.
// *
// * You should have received a copy of the GNU General Public License along
// * with PDFKeeper. If not, see <https://www.gnu.org/licenses/>.
// ****************************************************************************

using System;
using System.Diagnostics;
using System.Globalization;
using System.IO;

namespace PDFKeeper.Core.Application
{
    public class HelpFile
    {
        public enum Topic
        {
            Donate,
            License,
            SetupMultiUserDatabase,
            SetupSingleUserDatabase,
            ThirdPartyNotices,
            UsingPDFKeeper
        }

        public HelpFile()
        {
            var executingAssembly = new ExecutingAssembly();
            var productName = executingAssembly.ProductName;
            FullName = string.Concat(
                productName,
                CultureInfo.CurrentCulture.ToString(),
                ".chm");

            if (!File.Exists(FullName))
            {
                FullName = string.Concat(productName, ".en-US.chm");
            }
        }

        public string FullName { get; private set; }

        /// <summary>
        /// Gets the file name of a <see cref="Topic"/>.
        /// </summary>
        /// <param name="topic">The <see cref="Topic"/>.</param>
        /// <returns>The file name.</returns>
        public static string GetTopicFileName(Topic topic)
        {
            string topicFileName = null;
            switch (topic)
            {
                case Topic.Donate:
                    topicFileName = "Donate.html";
                    break;
                case Topic.License:
                    topicFileName = "COPYING.html";
                    break;
                case Topic.SetupMultiUserDatabase:
                    topicFileName = "Setup Multi-User Database.html";
                    break;
                case Topic.SetupSingleUserDatabase:
                    topicFileName = "Setup Single-User Database.html";
                    break;
                case Topic.ThirdPartyNotices:
                    topicFileName = "THIRD-PARTY-NOTICES.html";
                    break;
                case Topic.UsingPDFKeeper:
                    topicFileName = "Using PDFKeeper.html";
                    break;
            }
            
            return topicFileName;
        }

        /// <summary>
        /// Shows a help file <see cref="Topic"/> and waits for the help dialog to be closed by the
        /// user.
        /// </summary>
        /// <param name="topic">The <see cref="Topic"/>.</param>
        public void ShowHelp(Topic topic)
        {
            using (var process = new Process())
            {
                process.StartInfo.FileName = Path.Combine(
                    Environment.GetFolderPath(
                        Environment.SpecialFolder.Windows),
                    "hh.exe");
                process.StartInfo.Arguments = string.Concat(
                    "ms-its:",
                    FullName,
                    "::",
                    GetTopicFileName(topic));
                process.Start();
                process.WaitForExit();
            }
        }
    }
}
